### PR TITLE
feat(Site): Implement domain redirects for multiple domains for a Site

### PR DESCRIPTION
**Pull request by balamurali27**
- [x] API to set primary domain of Site
- [x]  API to get primary domain of Site
- [x] API to setup redirect when redirect_to_primary is enabled for a Site Domain

Depends on https://github.com/frappe/agent/pull/17

### CODE REVIEW

File: press/api/site.py [hunk 1]
```diff
@@ -574,6 +603,18 @@ def set_host_name(name, domain):
...
+@protected("Site")
+def redirect_to_primary_domain(name, domain):
+	frappe.get_doc("Site", name).redirect_to_primary_domain(domain)
+
+
+@frappe.whitelist()
+@protected("Site")
+def undo_redirect_to_primary_domain(name, domain):
+	frappe.get_doc("Site", name).undo_redirect_to_primary_domain(domain)
+
```
Comment (gavindsouza) [inline press/api/site.py]: We'll have to think of leaner API names than this 🤔

File: press/patches/v0_0_1/make_default_site_domain.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+# MIT License. See license.txt
+
+from __future__ import unicode_literals
+import frappe
+
+
+def execute():
+	frappe.reload_doc("press", "doctype", "site")
+	for site in frappe.db.get_all("Site", {"status": "Active"}, ["name"]):
+		frappe.get_doc("Site", site.name)._create_default_site_domain()
```
Comment (gavindsouza) [inline press/patches/v0_0_1/make_default_site_domain.py]:
```suggestion
	for site in frappe.db.get_all("Site", {"status": ("!=", "Archived")}, pluck="name"):
		frappe.get_doc("Site", site)._create_default_site_domain()
```

File: press/press/doctype/site/site.py [hunk 3]
```diff
@@ -52,13 +52,17 @@ def validate(self):
...
+					f"Frappe App {app.app} is not available on Bench {self.bench}."
+				)
 		frappe_app = self.apps[0]
 		if not frappe.db.get_value("Frappe App", frappe_app.app, "frappe"):
 			frappe.throw("First app to be installed on site must be frappe.")
 		apps = [app.app for app in self.apps]
 		if len(apps) != len(set(apps)):
 			frappe.throw("Can't install same app twice.")
+		if not self.is_new():
+			self._verify_host_name()
```
Comment (gavindsouza) [inline press/press/doctype/site/site.py]: Maybe don't run this check each time the site doc is updated. Use `get_doc_before_save` or `has_value_changed` to check if `host_name` has changed.

File: press/press/doctype/site/site.py [hunk 4]
```diff
@@ -126,16 +134,34 @@ def can_create_site(self):
...
+			}
+		).insert(ignore_if_duplicate=True)
+
 	def after_insert(self):
 		# create a site plan change log
 		self._create_initial_site_plan_change()
 		# log activity
 		log_site_activity(self.name, "Create")
 		self.create_agent_request()
+		self.set_host_name(self._create_default_site_domain().name)
```
Comment (gavindsouza) [inline press/press/doctype/site/site.py]: `_create_default_site_domain` can be changed to `_default_site_domain` as a property of the site. 
Use pluck for `name`

File: press/press/doctype/site/site.py [hunk 5]
```diff
@@ -251,10 +279,58 @@ def retry_add_domain(self, domain):
...
+		self.update_site_config({"host_name": f"https://{self.host_name}"})
+		self._update_redirects_for_all_site_domains()
+
+	def _update_redirects_for_all_site_domains(self):
+		site_domains = frappe.get_all(
+			"Site Domain", filters={"site": self.name, "redirect_to_primary": True},
+		)
+		domains = [domain["name"] for domain in site_domains]
+		for domain in domains:
+			self.redirect_to_primary_domain(domain)
```
Comment (gavindsouza) [inline press/press/doctype/site/site.py]:
```suggestion
		site_domains = frappe.get_all(
			"Site Domain", filters={"site": self.name, "redirect_to_primary": True}, pluck="name"
		)
		for domain in site_domains:
			self.redirect_to_primary_domain(domain)
```

File: press/press/doctype/site/site.js [hunk 6]
```diff
@@ -68,4 +76,7 @@ frappe.ui.form.on('Site', {
 			);
 		});
 	},
+	after_save: function(frm){
+		frm.call('update_host_name_in_agent');
+	}
```
Comment (gavindsouza) [inline press/press/doctype/site/site.js]: Move this in site.py in the validate event along with the `_verify_host_name` check

File: press/press/doctype/site_domain/site_domain.js [hunk 7]
```diff
@@ -2,7 +2,11 @@
...
-	// refresh: function(frm) {
-
-	// }
+	after_save: function (frm) {
+		if (frm.doc.redirect_to_primary) {
+			frm.call('setup_redirect');
+		} else {
+			frm.call('remove_redirect');
+		}
+	},
```
Comment (gavindsouza) [inline press/press/doctype/site_domain/site_domain.js]: Check if this can be done in the backend

File: press/press/doctype/site/test_site.py [hunk 8]
```diff
@@ -3,9 +3,49 @@
...
 
+import frappe
+
+from ..bench.test_bench import create_test_bench
+from ..frappe_app.test_frappe_app import create_test_frappe_app
+from ..plan.test_plan import create_test_plan
+from ..proxy_server.test_proxy_server import create_test_proxy_server
+from ..release_group.test_release_group import create_test_release_group
+from ..server.test_server import create_test_server
+from .site import Site
```
Comment (gavindsouza) [inline press/press/doctype/site/test_site.py]: Use absolute imports (as per our decision)

@adityahase @netchampfaris Share your thoughts on this

File: press/api/site.py (hunk 1 above)
Comment (balamurali27) [reply press/api/site.py]: how about this 4fbd939

File: press/patches/v0_0_1/make_default_site_domain.py (hunk 2 above)
Comment (balamurali27) [reply press/patches/v0_0_1/make_default_site_domain.py]: You meant to select all sites that aren't `Archived` here I believe

File: press/press/doctype/site/site.py (hunk 3 above)
Comment (balamurali27) [reply press/press/doctype/site/site.py]: Updated here 413bff4

File: press/patches/v0_0_1/make_default_site_domain.py (hunk 2 above)
Comment (balamurali27) [reply press/patches/v0_0_1/make_default_site_domain.py]: Updated here bc60cad

File: press/press/doctype/site/site.py (hunk 4 above)
Comment (balamurali27) [reply press/press/doctype/site/site.py]: Hey, the function is used only here to create the default site domain. The default site domain is not needed anywhere else. There's no need to make it a property, right?

File: press/press/doctype/site/site.py (hunk 5 above)
Comment (balamurali27) [reply press/press/doctype/site/site.py]: Updated in 6edfa00

File: press/press/doctype/site/site.js (hunk 6 above)
Comment (balamurali27) [reply press/press/doctype/site/site.js]: Updated in 43378ff

File: press/press/doctype/site_domain/site_domain.js (hunk 7 above)
Comment (balamurali27) [reply press/press/doctype/site_domain/site_domain.js]: Updated in d34143b

**Pull request merged**